src/Orleans/Messaging/SocketManager.cs

```diff
@@ -65,9 +67,16 @@ private Socket SendingSocketCreator(IPEndPoint target)
                 s.NoDelay = true;
                 WriteConnectionPreemble(s, Constants.SiloDirectConnectionId); // Identifies this client as a direct silo-to-silo socket
                 // Start an asynch receive off of the socket to detect closure
-                var foo = new byte[4];
-                s.BeginReceive(foo, 0, 1, SocketFlags.None, ReceiveCallback,
-                    new Tuple<Socket, IPEndPoint, SocketManager>(s, target, this));
+                var receiveAsyncEventArgs = new SocketAsyncEventArgs
+                {
+                    BufferList = new List<ArraySegment<byte>> { new ArraySegment<byte>(new byte[4]) },
+                    UserToken = new Tuple<Socket, IPEndPoint, SocketManager>(s, target, this)
+                };
+                receiveAsyncEventArgs.Completed += ReceiveCallback;
+                if (!s.ReceiveAsync(receiveAsyncEventArgs))
```

Comment:
I'd be interested to know of the performance implications (as [pondered](https://github.com/dotnet/orleans/issues/307#issuecomment-197551455)). While there, a post by *Toub* might be tangentially interest [here](http://blogs.msdn.com/b/pfxteam/archive/2011/12/15/10248293.aspx).